File: JavaScript/549-Binary Tree Longest Consecutive Sequence II.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var longestConsecutive = function(root) {
    let res = 0;
    
    let dfs = (node) =>{
        
        let inc = 0, dec = 0;
        
        if (node.left){
            let [dec1, inc1] = dfs(node.left);
            if (node.val + 1 === node.left.val){
                dec = dec1;
            }else if (node.val - 1 === node.left.val){
                inc = inc1;
            }
        }
        
        if (node.right){
            let [dec2, inc2] = dfs(node.right);
            if (node.val + 1 === node.right.val){
                dec = Math.max(dec, dec2);
            }else if (node.val - 1 === node.right.val){
                inc = Math.max(inc, inc2);
            }
        }
        res = Math.max(res, inc + 1 + dec);
        return [dec+1, inc+1];
    }
    dfs(root);
    return res;
};

File: C++/549-Binary Tree Longest Consecutive Sequence II.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int res = 1;
    pair<int, int> dfs(TreeNode* node) {
        int leftIncreasing = 0, leftDecreasing = 0, rightIncreasing = 0, rightDecreasing = 0;
        if (node->left != nullptr) {
            auto [inc, dec] = dfs(node->left);
            if (node->val == node->left->val + 1) {
                leftIncreasing = inc;
            }
            if (node->val == node->left->val - 1) {
                leftDecreasing = dec;
            }
        }
        if (node->right != nullptr) {
            auto [inc, dec] = dfs(node->right);
            if (node->val == node->right->val + 1) {
                rightIncreasing = inc;
            }
            if (node->val == node->right->val - 1) {
                rightDecreasing = dec;
            }
        }

        res = max({res, leftIncreasing + 1 + rightDecreasing, rightIncreasing + 1 + leftDecreasing});

        return {max(1+leftIncreasing, 1+rightIncreasing), max(1+leftDecreasing, 1+rightDecreasing)};
    }
    int longestConsecutive(TreeNode* root) {
        dfs(root);

        return res;
    }
};



// Better
class Solution {
public:
    int res = 1;
    pair<int, int> dfs(TreeNode* node) {
        int increasing = 0, decreasing = 0;
        if (node->left != nullptr) {
            auto [inc, dec] = dfs(node->left);
            if (node->val == node->left->val + 1) {
                increasing = inc;
            }
            if (node->val == node->left->val - 1) {
                decreasing = dec;
            }
        }
        if (node->right != nullptr) {
            auto [inc, dec] = dfs(node->right);
            if (node->val == node->right->val + 1) {
                increasing = max(increasing, inc);
            }
            if (node->val == node->right->val - 1) {
                decreasing = max(decreasing, dec);
            }
        }

        res = max(res, increasing + 1 + decreasing);

        return {1+increasing, 1+decreasing};
    }
    int longestConsecutive(TreeNode* root) {
        dfs(root);

        return res;
    }
};